### PR TITLE
Improve compact encoding/decoding

### PR DESCRIPTION
**Pull request by jprupp**
Compact numbers can also be negative. Code rewritten for better compatibility with reference implementation.

### CODE REVIEW

**Comment by m0dev76** [comment]
I guess the changes were to reflect the code in bitcoin core?

https://github.com/bitcoin/bitcoin/blob/master/src/arith_uint256.cpp#L206

File: haskoin-core/src/Network/Haskoin/Block/Types.hs [hunk 1]
```diff
@@ -339,28 +335,51 @@ instance Serialize Headers where
...
-    c1 | s1 < 3    = posi `shiftL` (8*(3 - s1))
-       | otherwise = posi `shiftR` (8*(s1 - 3))
-    (s2,c2) | c1 .&. 0x00800000 /= 0  = (s1 + 1, c1 `shiftR` 8)
-            | otherwise               = (s1, c1)
-    c3 = fromIntegral $ c2 .|. (toInteger s2 `shiftL` 24)
-
+    i' = abs i
+    neg = i < 0
+    nSize' :: Int
+    nSize' = let f n = let n' = n `shiftR` 8
```
Comment (m0dev76) [inline haskoin-core/src/Network/Haskoin/Block/Types.hs]: According to https://github.com/bitcoin/bitcoin/blob/master/src/arith_uint256.cpp#L228 `nSize == 0` is a valid possibility but this code will always give at least `1`. Maybe consider something like this:

```
nSize' = let f n = if n == 0 then 0 else 1 + f (n `shiftR` 8)
          in f i' 
```

File: haskoin-core/test/Network/Haskoin/Crypto/Hash/Units.hs [hunk 2]
```diff
@@ -40,31 +36,22 @@ mapDRBGRsd vs = testCase "HMAC DRBG Vectors" $ mapM_ testDRBGRsd $ zip vs [0..]
...
-    let w1 = hmacDRBGNew (v !! 0) (v !! 1) (v !! 2)
+    let w1 = hmacDRBGNew (head v) (v !! 1) (v !! 2)
         w2 = hmacDRBGRsd w1 (v !! 3) (v !! 4)
         (w3,_) = hmacDRBGGen w2 128 (v !! 5)
         (_,r)  = hmacDRBGGen w3 128 (v !! 6)
     assertBool name $ fromJust r == (v !! 7)
     where v = map (fromJust . decodeHex) s
-          name = "    > HMAC DRBG Vector " ++ (show i)
-
-testCompact :: Assertion
```
Comment (m0dev76) [inline haskoin-core/test/Network/Haskoin/Crypto/Hash/Units.hs:60]: Why remove these tests? Are they now invalid? If so maybe replace them with new valid ones?

**Review by m0dev76 — CHANGES_REQUESTED** [review]

File: haskoin-core/src/Network/Haskoin/Block/Types.hs (hunk 1 above)
Comment (jprupp) [reply haskoin-core/src/Network/Haskoin/Block/Types.hs]: That was an excellent catch sir.

Image of a big fish

I added new test vectors that fail with the buggy code, as well as a fix for this issue.

File: haskoin-core/test/Network/Haskoin/Crypto/Hash/Units.hs (hunk 2 above)
Comment (jprupp) [reply haskoin-core/test/Network/Haskoin/Crypto/Hash/Units.hs:60]: Tests have been moved. They still exist.

**Review by m0dev76 — APPROVED** [review]

**Pull request merged**